File: clojure/bob/src/bob.clj
(ns bob)

(defn silence? [s]
  (re-find #"^\s*$" s))

(defn yelling? [s]
  (and (re-find #"[A-Z]" s)
       (re-find #"^[^a-z]+$" s)))

(defn question? [s]
  (re-find #"\?$" s))

(defn yelling-question? [s]
  (and (yelling? s)
       (question? s)))

(defn response-for [s]
  (cond
    (silence? s)          "Fine. Be that way!"
    (yelling-question? s) "Calm down, I know what I'm doing!"
    (yelling? s)          "Whoa, chill out!"
    (question? s)         "Sure."
    :else                 "Whatever."))

;;
;; If you don't like the repetition of `s' you could use an anonymous fn
;;
;; (defn response-for [s]
;;   (condp #(% %2) s
;;     silence?          "Fine. Be that way!"
;;     yelling-question? "Calm down, I know what I'm doing!"
;;     yelling?          "Whoa, chill out!"
;;     question?         "Sure."
;;     "Whatever."))

File: elixir/bob/bob.exs
defmodule Bob do
  def hey(input) do
    cond do
      silence?(input) -> "Fine. Be that way!"
      yelling_a_question?(input) -> "Calm down, I know what I'm doing!"
      yelling?(input) -> "Whoa, chill out!"
      question?(input) -> "Sure."
      true -> "Whatever." 
    end
  end

  def yelling?(input) do
    case String.match?(input, ~r/\p{L}/) do
      true ->  yelling_alpha?(input)
      false -> false
    end
  end

  def yelling_alpha?(input) do
    input == String.upcase(input)
  end

  def question?(input) do
    String.ends_with?(input, "?")
  end

  def yelling_a_question?(input) do
    yelling?(input) and question?(input)
  end

  def silence?(""), do: true 
  def silence?(input) do
    String.match?(input, ~r/^\s/)
  end
end
